// @composio/client/resources/tool-router/session.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Session extends APIResource {
    /**
     * Creates a new session for the tool router feature. This endpoint initializes a
     * new session with specified toolkits and their authentication configurations. The
     * session provides an isolated environment for testing and managing tool routing
     * logic with scoped MCP server access.
     *
     * @example
     * ```ts
     * const session = await client.toolRouter.session.create({
     *   user_id: 'user_123456789',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/v3/tool_router/session', { body, ...options });
    }
    /**
     * Retrieves an existing tool router session by its ID. Returns the session
     * configuration, MCP server URL, and available tools.
     *
     * @example
     * ```ts
     * const session = await client.toolRouter.session.retrieve(
     *   'trs_123456789',
     * );
     * ```
     */
    retrieve(sessionID, options) {
        return this._client.get(path `/api/v3/tool_router/session/${sessionID}`, options);
    }
    /**
     * Executes a specific tool within a tool router session. The toolkit is
     * automatically inferred from the tool slug. The tool must belong to an allowed
     * toolkit and must not be disabled in the session configuration. This endpoint
     * validates permissions, resolves connected accounts, and executes the tool with
     * the session context.
     *
     * @example
     * ```ts
     * const response = await client.toolRouter.session.execute(
     *   'trs_LX9uJKBinWWr',
     *   { tool_slug: 'GITHUB_CREATE_ISSUE' },
     * );
     * ```
     */
    execute(sessionID, body, options) {
        return this._client.post(path `/api/v3/tool_router/session/${sessionID}/execute`, { body, ...options });
    }
    /**
     * Executes a Composio meta tool (COMPOSIO\_\*) within a tool router session.
     *
     * @example
     * ```ts
     * const response =
     *   await client.toolRouter.session.executeMeta(
     *     'trs_LX9uJKBinWWr',
     *     { slug: 'COMPOSIO_MANAGE_CONNECTIONS' },
     *   );
     * ```
     */
    executeMeta(sessionID, body, options) {
        return this._client.post(path `/api/v3/tool_router/session/${sessionID}/execute_meta`, {
            body,
            ...options,
        });
    }
    /**
     * Initiates an authentication link session for a specific toolkit within a tool
     * router session. Returns a link token and redirect URL that users can use to
     * complete the OAuth flow.
     *
     * @example
     * ```ts
     * const response = await client.toolRouter.session.link(
     *   'trs_LX9uJKBinWWr',
     *   { toolkit: 'github' },
     * );
     * ```
     */
    link(sessionID, body, options) {
        return this._client.post(path `/api/v3/tool_router/session/${sessionID}/link`, { body, ...options });
    }
    /**
     * Retrieves a cursor-paginated list of toolkits available in the tool router
     * session. Includes toolkit metadata, composio-managed auth schemes, and connected
     * accounts if available. Optionally filter by specific toolkit slugs.
     *
     * @example
     * ```ts
     * const response = await client.toolRouter.session.toolkits(
     *   'trs_123456789',
     * );
     * ```
     */
    toolkits(sessionID, query = {}, options) {
        return this._client.get(path `/api/v3/tool_router/session/${sessionID}/toolkits`, { query, ...options });
    }
    /**
     * Returns the meta tools available in a tool router session with their complete
     * schemas. This includes request and response schemas specific to the session
     * context.
     *
     * @example
     * ```ts
     * const response = await client.toolRouter.session.tools(
     *   'session_id',
     * );
     * ```
     */
    tools(sessionID, query, options) {
        return this._client.get(path `/api/v3/tool_router/session/${sessionID}/tools`, options);
    }
}
