// @composio/client/resources/auth-configs.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthConfigs = void 0;
const resource_1 = require("../core/resource.js");
const path_1 = require("../internal/utils/path.js");
class AuthConfigs extends resource_1.APIResource {
    /**
     * Create new authentication configuration
     *
     * @example
     * ```ts
     * const authConfig = await client.authConfigs.create({
     *   toolkit: { slug: 'slug' },
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/api/v3/auth_configs', { body, ...options });
    }
    /**
     * Retrieves detailed information about a specific authentication configuration
     * using its unique identifier.
     *
     * @example
     * ```ts
     * const authConfig = await client.authConfigs.retrieve(
     *   'nanoid',
     * );
     * ```
     */
    retrieve(nanoid, options) {
        return this._client.get((0, path_1.path) `/api/v3/auth_configs/${nanoid}`, options);
    }
    /**
     * Modifies an existing authentication configuration with new credentials or other
     * settings. Only specified fields will be updated.
     *
     * @example
     * ```ts
     * const authConfig = await client.authConfigs.update(
     *   'nanoid',
     *   { type: 'custom' },
     * );
     * ```
     */
    update(nanoid, body, options) {
        return this._client.patch((0, path_1.path) `/api/v3/auth_configs/${nanoid}`, { body, ...options });
    }
    /**
     * List authentication configurations with optional filters
     *
     * @example
     * ```ts
     * const authConfigs = await client.authConfigs.list();
     * ```
     */
    list(query = {}, options) {
        return this._client.get('/api/v3/auth_configs', { query, ...options });
    }
    /**
     * Soft-deletes an authentication configuration by marking it as deleted in the
     * database. This operation cannot be undone.
     *
     * @example
     * ```ts
     * const authConfig = await client.authConfigs.delete(
     *   'nanoid',
     * );
     * ```
     */
    delete(nanoid, options) {
        return this._client.delete((0, path_1.path) `/api/v3/auth_configs/${nanoid}`, options);
    }
    /**
     * Updates the status of an authentication configuration to either enabled or
     * disabled. Disabled configurations cannot be used for new connections.
     *
     * @example
     * ```ts
     * const response = await client.authConfigs.updateStatus(
     *   'ENABLED',
     *   { nanoid: 'nanoid' },
     * );
     * ```
     */
    updateStatus(status, params, options) {
        const { nanoid } = params;
        return this._client.patch((0, path_1.path) `/api/v3/auth_configs/${nanoid}/${status}`, options);
    }
}
exports.AuthConfigs = AuthConfigs;
