// @composio/client/src/resources/project/config.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import { APIPromise } from '../../core/api-promise';
import { RequestOptions } from '../../internal/request-options';

export class Config extends APIResource {
  /**
   * Retrieves the current project configuration including 2FA settings.
   *
   * @example
   * ```ts
   * const config = await client.project.config.retrieve();
   * ```
   */
  retrieve(options?: RequestOptions): APIPromise<ConfigRetrieveResponse> {
    return this._client.get('/api/v3/org/project/config', options);
  }

  /**
   * Updates the project configuration settings.
   *
   * @example
   * ```ts
   * const config = await client.project.config.update({
   *   is_2FA_enabled: true,
   * });
   * ```
   */
  update(
    body: ConfigUpdateParams | null | undefined = {},
    options?: RequestOptions,
  ): APIPromise<ConfigUpdateResponse> {
    return this._client.patch('/api/v3/org/project/config', { body, ...options });
  }
}

export interface ConfigRetrieveResponse {
  is_2FA_enabled: boolean;

  log_visibility_setting: 'show_all' | 'dont_store_data';

  mask_secret_keys_in_connected_account: boolean;

  display_name?: string;

  /**
   * Whether to enable composio link for managed authentication. This key will be
   * deprecated in the future. Please don't use this key.
   */
  is_composio_link_enabled_for_managed_auth?: boolean;

  logo_url?: string;

  require_mcp_api_key?: boolean;

  signed_url_file_expiry_in_seconds?: number;
}

export interface ConfigUpdateResponse {
  is_2FA_enabled: boolean;

  log_visibility_setting: 'show_all' | 'dont_store_data';

  mask_secret_keys_in_connected_account: boolean;

  display_name?: string;

  /**
   * Whether to enable composio link for managed authentication. This key will be
   * deprecated in the future. Please don't use this key.
   */
  is_composio_link_enabled_for_managed_auth?: boolean;

  logo_url?: string;

  require_mcp_api_key?: boolean;

  signed_url_file_expiry_in_seconds?: number;
}

export interface ConfigUpdateParams {
  display_name?: string;

  is_2FA_enabled?: boolean;

  /**
   * Whether to enable composio link for managed authentication. This key will be
   * deprecated in the future. Please don't use this key.
   */
  is_composio_link_enabled_for_managed_auth?: boolean;

  log_visibility_setting?: 'show_all' | 'dont_store_data';

  logo_url?: string;

  mask_secret_keys_in_connected_account?: boolean;

  require_mcp_api_key?: boolean;

  signed_url_file_expiry_in_seconds?: number;
}

export declare namespace Config {
  export {
    type ConfigRetrieveResponse as ConfigRetrieveResponse,
    type ConfigUpdateResponse as ConfigUpdateResponse,
    type ConfigUpdateParams as ConfigUpdateParams,
  };
}
